SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/InterestManager.cpp

```diff
@@ -0,0 +1,647 @@
+	{
+		CachedServerInterest[0].SetNum(NumRegions);
+		CachedServerInterest[1].SetNum(NumRegions);
+		CachedServerInterest[2].SetNum(NumRegions);
+	}
+
+	constexpr bool bUseBroadphaseForRegions = false;
+
+	if (Broadphase && bUseBroadphaseForRegions)
+	{
+		// 60ms, 200k entities, 64 regions, 5% overlap
+		// Worse than the non-spatialized version, which is not surprising, since we are visiting all the entities anyway,
+		// hence visiting the tree before testing the entities is just more data to process.
+		// Have to test it for the player interest since the tree culling might actually do something in that case.
+
+		// Since the regions are a partition of the entire space in the server case, a better strategy would be to find the region in the
+		// partition an entity belongs to and test the position against the segment boundaries with the interest border. The region's
+		// connectivity would be needed to interpret these results.
+		SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputationBroadphase);
+		const uint32 NumActiveEntities = ActiveEntities;
+
+		uint64* VisibilityPtr = Visibility.GetData();
+		const uint32* Flags = EntityFlags.GetData();
+		for (uint32 i = 0; i < NumActiveEntities; ++i)
+		{
+			const uint64 Mask = UINT64_MAX;
+			*VisibilityPtr = ((uint64)(*Flags != 0)) * Mask;
+			++VisibilityPtr;
+			++Flags;
+		}
+
+		Broadphase->ComputeBoxVisibility(Regions, Visibility.GetData());
+	}
+
+	constexpr bool bVectorizeRegions = true;
+	constexpr bool bVectorizeEntities = false;
+
+	if (!bVectorizeRegions && !bVectorizeEntities)
+	{
+		// Not-so naively iterate over the region
+		// 24ms, 200k entities, 64 regions
+		SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputationBox);
+		const FVector2D* Pos = EntityPosition.GetData();
+		const uint32* Flags = EntityFlags.GetData();
+		uint64* VisibilityPtr = Visibility.GetData();
+
+		const uint32 NumActiveEntities = ActiveEntities;
+
+		for (uint32 i = 0; i < NumActiveEntities; ++i)
+		{
+			const uint64 AllRegionsMask = UINT64_MAX;
+			*VisibilityPtr = ((uint64)(*Flags != 0)) * AllRegionsMask;
+
+			// Avoid using operator[] as the range-check could expand to a sizable chunk of code.
+			const FBox2D* RegionBox = Regions.GetData();
+
+			for (uint32 j = 0; j < NumRegions; ++j)
+			{
+				const uint64 Mask = 1ull << j;
+				// Setting individual bits is useful when doing a batch test over several regions in one go
+				*VisibilityPtr |= ((uint64)RegionBox->IsInside(*Pos)) * Mask;
+				++RegionBox;
+			}
+			++Pos;
+			++Flags;
+			++VisibilityPtr;
+		}
+	}
+
+	const uint64 AllRegionsMask = [NumRegions] {
+		// To handle the 64 regions case, shifting by 64 could be a nop;
+		uint64 Mask = 1ull << (NumRegions - 1);
+		Mask <<= 1;
+		return Mask - 1ull;
+	}();
+
+	if (bVectorizeRegions)
+	{
+		SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputationBoxSSE);
+
+		// SSE version, processing 4 regions at a time.
+		// 6ms, 200k entities, 64 regions
+		const uint32 Num4Regions = (NumRegions + 3) / 4;
```

Comment:
this potentially truncates is that what we want? worth calling out if intentional?

Reply:
Yeah truncates, but that's why we add 3 to it first.

Reply:
minor: but would prefer to use an explicit rounding function for clarity.